### PR TITLE
Have namedtuple `__replace__` return `Self`

### PR DESCRIPTION
**Pull request by max-muoto**
Have `__replace__` for named tuples return `Self` to ensure protocol compatibility with `copy.replace` in the typeshed, and to be accurate to the actual dunder.

### CODE REVIEW

File: mypy/semanal_namedtuple.py [hunk 1]
```diff
@@ -633,7 +634,7 @@ def add_method(
         if self.options.python_version >= (3, 13):
             add_method(
                 "__replace__",
-                ret=None,
+                ret=Instance(info, []),
```
Comment (JelleZijlstra) [inline mypy/semanal_namedtuple.py:637]: This also needs to deal with generics.

File: test-data/unit/check-namedtuple.test [hunk 2]
```diff
@@ -1407,9 +1407,23 @@ from typing import NamedTuple
...
+from typing import TypeVar, Generic
+
+T = TypeVar("T")
+
+class GenericA(NamedTuple, Generic[T]):
+    x: T
+
+replaced_2 = GenericA(x=0).__replace__(x=1)
+reveal_type(replaced_2)  # N: Revealed type is "__main__.GenericA"
+GenericA(x=0).__replace__(x="abc") # E: Argument "x" to "__replace__" of "GenericA" has incompatible type "str"; expected "int"
```
Comment (max-muoto) [inline test-data/unit/check-namedtuple.test:1425]: It's specialized, but the generic isn't in the revealed type which is odd.

Comment (JelleZijlstra) [reply test-data/unit/check-namedtuple.test:1425]: I think that's how mypy generally works:

```
% mypy -c 'x: list[str] = []; x.append(1)'
<string>:1: error: Argument 1 to "append" of "list" has incompatible type "int"; expected "str"  [arg-type]
```

Agree this is potentially confusing but it's not something that should be changed in this PR.

Comment (max-muoto) [reply test-data/unit/check-namedtuple.test:1425]: Ah I see, makes sense - let me bring this out of draft in that case.

**Comment by ilevkivskyi** [comment]
I am sorry but I am going to revert this. Even if it does fix something, this is a very wrong way to do fix it.

**Comment by max-muoto** [comment]
> I am sorry but I am going to revert this. Even if it does fix something, this is a very wrong way to do fix it.

Apologies there. How would you recommend going about fixing this in that case?

**Comment by ilevkivskyi** [comment]
> How would you recommend going about fixing this in that case?

For starter, it would be great to open an issue with a (detailed) explanation of what is wrong.

**Comment by max-muoto** [comment]
> > How would you recommend going about fixing this in that case?
> 
> For starter, it would be great to open an issue with a (detailed) explanation of what is wrong.

Sure thing - in short summary it replaces to how `copy.replace` is handled in 3.13, and the need for `__replace__` to match a protocol which implements `def __replace(self, *args: Any, **kwargs: Any) -> Self: ...` but let me write something up (think [this issue](https://github.com/python/mypy/issues/17471), but for named-tuples).

https://github.com/python/mypy/issues/17497

**Pull request merged**